daml/ContingentClaims/Lifecycle.daml

```diff
@@ -3,83 +3,120 @@
+--           \----------------------- monad ------------------------/  \-------- value --------/
 
-lifecycle' : (Eq a, Monad m, CanAbort m)
+-- | HIDE Helper function for sequencing algebras.
+distM : Monad m
+      => Either (C a) (WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (C a)))
+      -> M m a
+distM (Left c) = pure (Left <$> (project c))
+distM (Right m) = fmap Right <$> m
+
+-- | Evaluate observables and skip branches for which predicates don't hold.
+-- This is useless onl it's own. Composed with other functions it adds laziness.
+lazily' : (Eq a, Monad m)
   => (Text -> Date -> m Decimal)
-  -> (C a -> C a -> C a)
-  -> C a -> Seed m a
+  -> C a -> Kleisli m Date (F a (Either (C a) (C a)))
+lazily' spot (When obs c) = do
+  predicate <- eval spot obs
+  pure if predicate then WhenF obs (Right c) else WhenF obs (Left c)
+lazily' spot (Cond obs c c') = do
+  predicate <- eval spot obs
+  if predicate then lazily' spot c else lazily' spot c'
+lazily' _ other = pure $ Right <$> project other
 
-lifecycle' _ _ Zero = pure ZeroF
-lifecycle' _ _ (One id) = do
+-- | Collects all `One` leaves replacing them with `Zero`s (to avoid double-spend).
+-- Calculates the product of any `Scale` factors between each `One` and the root node, logging these to the writer.
+-- This function **ignores** conditions; it traverses the entire tree.
+settle' : CanAbort m => (Text -> Date -> m Decimal) -> C a -> WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (C a))
+settle' _ (One asset) = do
   qty <- get
-  tell [(qty, id)]
-  pure ZeroF -- replace lifecycled node with zero, to avoid double-spend.
-lifecycle' _ _ (Give c) = do
-  modify negate 
-  pure $ GiveF (Right c)
-lifecycle' spot _ (When obs c) = do
-  predicate <- lift . lift $ eval spot obs
-  pure $ if predicate then WhenF obs (Right c) else WhenF obs (Left c)
-lifecycle' spot _ (Scale obs c) = do
+  tell [(qty, asset)]
+  pure ZeroF
+settle' _ (Give c) = do
+  modify negate
+  pure $ GiveF c
+settle' spot (Scale obs c) = do
   k <- lift . lift $ eval spot obs
-  modify (* k) 
-  pure $ ScaleF obs (Right c)
-lifecycle' spot choose (Or c c') = do
-  let chosen = choose c c'
-  if | chosen == c -> lifecycle' spot choose c --TODO: should we abort with nested choices?
-     | chosen == c' -> lifecycle' spot choose c'
-     | otherwise -> abort "lifecycle: Invalid branch election"
-lifecycle' _ _ (And c c') = pure $ AndF (Right c) (Right c')
-lifecycle' spot choose (Cond obs c c') = do
-  predicate <- lift . lift $ eval spot obs
-  if predicate then lifecycle' spot choose c else lifecycle' spot choose c'
+  modify (* k)
+  pure $ ScaleF obs c
+settle' _ (Or c c') = abort "settle: unexercised `Or`. Exercise it and then retry. "
+settle' _ other = pure $ project other
+
+-- | Used to make an election on `Or` or exercise an `Anytime`.
+--  `import` this `qualified`, to avoid clashes with `Prelude.exercise`.
```

Comment:
I'm aware of the name clash, but this is unfortunately the best term available. I considered `elect` also.

I don't see any problem using it qualified e.g. `Lifecycle.exercise`.